Tests/ASTPreprocessorTests/ASTPreprocessorTests.swift

```diff
@@ -0,0 +1,151 @@
+    // Given
+    let f = Fixture()
+
+    // a.(b.c)
+    let originalTree =
+      BinaryExpression(
+        lhs: namedIdentifier("a"),
+        op: dotToken,
+        rhs: .binaryExpression(
+          BinaryExpression(
+            lhs: namedIdentifier("b"),
+            op: dotToken,
+            rhs: namedIdentifier("c"))))
+
+    // (a.b).c
+    let requiredTree =
+      BinaryExpression(
+        lhs: .binaryExpression(
+          BinaryExpression(
+            lhs: namedIdentifier("a"),
+            op: dotToken,
+            rhs: namedIdentifier("b"))),
+        op: dotToken,
+        rhs: namedIdentifier("c"))
+
+    // When
+    let result = f.pass.process(binaryExpression: originalTree, passContext: ASTPassContext())
+
+    // Then
+    XCTAssertTrue(result.element == requiredTree)
+  }
+
+  func testBinaryExpression_doesNotNeedLeftRotation_leftRotationNotApplied() {
+    // Given
+    let f = Fixture()
+
+    // (a.b).c
+    let tree =
+      BinaryExpression(
+        lhs: .binaryExpression(
+          BinaryExpression(
+            lhs: namedIdentifier("a"),
+            op: dotToken,
+            rhs: namedIdentifier("b"))),
+        op: dotToken,
+        rhs: namedIdentifier("c"))
+
+    // When
+    let result = f.pass.process(binaryExpression: tree, passContext: ASTPassContext())
+
+    // Then
+    XCTAssertTrue(result.element == tree)
+  }
+
+  func testBinaryExpression_childrenNotTreelike_leftRotationNotApplied() {
+    // Given
+    let f = Fixture()
+
+    // a.(b.c)
+    // This expression looks similar to those above, but this is an explicit bracketing that we need to allow
+    let tree =
+      BinaryExpression(
+        lhs: namedIdentifier("a"),
+        op: dotToken,
+        rhs: .bracketedExpression(
+          BracketedExpression(
+            expression: .binaryExpression(
+              BinaryExpression(
+                lhs: namedIdentifier("b"),
+                op: dotToken,
+                rhs: namedIdentifier("c"))),
+            openBracketToken: Token(kind: .punctuation(.openBracket), sourceLocation: .DUMMY),
+            closeBracketToken: Token(kind: .punctuation(.closeBracket), sourceLocation: .DUMMY))))
+
+    // When
+    let result = f.pass.process(binaryExpression: tree, passContext: ASTPassContext())
+
+    // Then
+    XCTAssertTrue(result.element == tree)
+  }
+
+  func testBinaryExpression_notDotKind_leftRotationNotApplied() {
+    // Given
+    let f = Fixture()
+
+    // a + b
+    let tree =
+      BinaryExpression(
+        lhs: namedIdentifier("a"),
+        op: Token(kind: .punctuation(.plus), sourceLocation: .DUMMY),
+        rhs: namedIdentifier("b"))
+
+    // When
+    let result = f.pass.process(binaryExpression: tree, passContext: ASTPassContext())
+
+    // Then
+    XCTAssertTrue(result.element == tree)
+  }
+
+  func testBinaryExpression_rightChildNotDotKind_leftRotationNotApplied() {
+    // Given
+    let f = Fixture()
+
+    // a.(b+c)
```

Comment:
Again, not valid code…

Reply:
Again, it doesn't matter whether the code is valid since we're not testing for validity here